torch/optim/asgd.py

```diff
@@ -62,11 +73,17 @@ def step(self, closure=None):
                 state['step'] += 1
 
                 if group['weight_decay'] != 0:
-                    grad = grad.add(group['weight_decay'], p.data)
+                    if group['l2_reg']:
+                        grad.add_(group['weight_decay'], p.data)
+                    else:
+                        xold = p.data.clone()
 
                 # decay term
                 p.data.mul_(1 - group['lambd'] * state['eta'])
 
+                if group['weight_decay'] != 0 and not group['l2_reg']:
+                    p.data.add_(-group['weight_decay'], xold)
```

Comment:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], xold)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).